src/java.base/share/classes/java/math/BigDecimal.java

```diff
@@ -3776,6 +3780,100 @@ public byte byteValueExact() {
+     * converted to {@link Float16#NEGATIVE_INFINITY} or {@link
+     * Float16#POSITIVE_INFINITY} as appropriate.  Note that even when
+     * the return value is finite, this conversion can lose
+     * information about the precision of the {@code BigDecimal}
+     * value.Float16
+     *
+     * @return this {@code BigDecimal} converted to a {@code Float16}.
+     * @jls 5.1.3 Narrowing Primitive Conversion
+     */
+    public Float16 float16Value() {
+        /* For details, see the extensive comments in doubleValue(). */
+        if (intCompact != INFLATED) {
+            Float16 v = Float16.valueOf(intCompact);
+            if (scale == 0) {
+                return v;
+            }
+            /*
+             * The discussion for the double case also applies here. That is,
+             * the following test is precise for all long values, but here
+             * Long.MAX_VALUE is not an issue.
+             */
+            if (v.longValue() == intCompact) {
+                if (0 < scale && scale < FLOAT16_10_POW.length) {
+                    return Float16.divide(v, FLOAT16_10_POW[scale]);
+                }
+                if (0 > scale && scale > -FLOAT16_10_POW.length) {
+                    return Float16.multiply(v, FLOAT16_10_POW[-scale]);
+                }
+            }
+        }
+        return fullFloat16Value();
+    }
+
+    private Float16 fullFloat16Value() {
+        if (intCompact == 0) {
+            return Float16.valueOf(0);
+        }
+        BigInteger w = unscaledValue().abs();
+        long qb = w.bitLength() - (long) Math.ceil(scale * L);
+        Float16 signum = Float16.valueOf(signum());
+        if (qb < Q_MIN_F16 - 2) {  // qb < -26
+            return Float16.multiply(signum, Float16.valueOf(0));
+        }
+        if (qb > Q_MAX_F16 + P_F16 + 1) {  // qb > 17
+            return Float16.multiply(signum, Float16.POSITIVE_INFINITY);
+        }
+        if (scale < 0) {
+            return Float16.multiply(signum, w.multiply(bigTenToThe(-scale)).float16Value());
+        }
+        if (scale == 0) {
+            return Float16.multiply(signum, w.float16Value());
+        }
+        int ql = (int) qb - (P_F16 + 3);
+        BigInteger pow10 = bigTenToThe(scale);
+        BigInteger m, n;
+        if (ql <= 0) {
+            m = w.shiftLeft(-ql);
+            n = pow10;
+        } else {
+            m = w;
+            n = pow10.shiftLeft(ql);
+        }
+        BigInteger[] qr = m.divideAndRemainder(n);
+        /*
+         * We have
+         *      2^12 = 2^{P+1} <= i < 2^{P+5} = 2^16
+         * Contrary to the double and float cases, where we use long and int, resp.,
+         * here we cannot simply declare i as short, because P + 5 < Short.SIZE
+         * fails to hold.
+         * Using int is safe, though.
+         *
+         * Further, as Math.scalb(Float16) does not exists, we fall back to
+         * Math.scalb(double).
+         */
```

Comment:
FTR @rgiulietti , Thanks for the pointers, comments looks out of context unless read in light of https://github.com/openjdk/valhalla/blob/lworld/src/java.base/share/classes/java/math/BigDecimal.java#L4066

We should add a reference to fullDoubleValue just before your explanation.